server/src/main/java/org/opensearch/index/RemoteRefreshSegmentPressureService.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.core.concurrency.OpenSearchRejectedExecutionException;
+import org.opensearch.index.shard.IndexEventListener;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.shard.ShardId;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Service used to validate if the incoming indexing request should be rejected based on the {@link RemoteRefreshSegmentPressureTracker}.
+ */
+public class RemoteRefreshSegmentPressureService implements IndexEventListener {
+
+    private static final Logger logger = LogManager.getLogger(RemoteRefreshSegmentPressureService.class);
+
+    /**
+     * Keeps map of remote-backed index shards and their corresponding backpressure tracker.
+     */
+    private final Map<ShardId, RemoteRefreshSegmentPressureTracker> trackerMap = ConcurrentCollections.newConcurrentMap();
+
+    /**
+     * Remote refresh segment pressure settings which is used for creation of the backpressure tracker and as well as rejection.
+     */
+    private final RemoteRefreshSegmentPressureSettings pressureSettings;
+
+    @Inject
+    public RemoteRefreshSegmentPressureService(ClusterService clusterService, Settings settings) {
+        pressureSettings = new RemoteRefreshSegmentPressureSettings(clusterService, settings, this);
```

Comment:
since we are only initializing this RemoteRefreshSegmentPressureSettings here it would always we single instance per process i guess? but in the long term we should actually modify RemoteRefreshSegmentPressureSettings itself to now allow multiple instantiations in single process. We can maybe do this later in another PR

Reply:
By multiple instantiation do you mean to make pressure settings as index level?